charabia/src/segmenter/camel_case.rs

```diff
@@ -0,0 +1,92 @@
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+pub(crate) trait CamelCaseSegmentation {
+    /// Returns an iterator over substrings of `self` separated on camelCase boundaries.
+    /// For instance, "camelCase" is split into ["camel", "Case"].
+    /// A camelCase boundary constitutes a lowercase letter directly followed by an uppercase letter
+    /// where lower and uppercase letters are defined by the corresponding Unicode General Categories.
+    fn split_camel_case_bounds(&self) -> CamelCaseParts;
+}
+
+pub(crate) struct CamelCaseParts<'t> {
+    state: State<'t>,
+}
+
+enum State<'t> {
+    InProgress { remainder: &'t str },
+    Exhausted,
+}
+
+impl CamelCaseSegmentation for str {
+    fn split_camel_case_bounds(&self) -> CamelCaseParts {
+        CamelCaseParts { state: State::InProgress { remainder: self } }
+    }
+}
+
+/// Matches a lowercase letter followed by an uppercase one with any number of optional non-spacing marks in between.
+static CAMEL_CASE_BOUNDARY_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"\p{Ll}\p{Mn}*?\p{Lu}").unwrap());
+
+impl<'t> Iterator for CamelCaseParts<'t> {
+    type Item = &'t str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.state {
+            State::Exhausted => None,
+            State::InProgress { remainder } => {
+                // CamelCase boundary consists of at least 2 code-points. Avoid expensive regex evaluation on shorter strings.
+                // Note that using `remainder.chars().count() == 1` may catch more cases (non-ASCII strings)
+                // but the main focus here is on " ", "-" and similar that are abundantly produced
+                // by `split_word_bounds()` in the Latin segmenter and mere `len()` performs better at that.
+                if remainder.len() == 1 {
+                    self.state = State::Exhausted;
+                    return Some(remainder);
+                }
+
+                match CAMEL_CASE_BOUNDARY_REGEX.find(remainder) {
```

Comment:
I was expecting a suggestion to avoid Regex but not a complete working example! 🖤

Turns out, this yields a performance boost of about 5%. I submitted a [PR](https://github.com/meilisearch/charabia/pull/196) with this change. Hopefully the maintainers won't be mad that we nerd out in their repo. 😅

One question: What's the reason for explicit `move` on the closure? The compiler doesn't complain about its absence.

Reply:
> One question: What's the reason for explicit `move` on the closure? The compiler doesn't complain about its absence.

I think I just copy-pasted it from a similar example, no reason to have it if compiler doesn't complain.

Without the `move`, the "desugared closure struct" (closures can be imagined as auto-generated structs with a call method) should look like
```rust
struct Closure {
    last_char_was_lower_case: bool  # mut is inherited, see https://stackoverflow.com/a/27023282/4345715
}
```
Without the `move` it would look like
```rust
struct Closure<'a> {
    last_char_was_lower_case: &'a mut bool
}
```

I.e. with the `move`, this closure becomes fully owned (and can be i.e. returned from function).

Without the `move`, the lifetime of the closure is limited by the `last_char_was_lower_case`'s lifetime. But that's apparently sufficient here.